File: scripts/windlogic.py
"""Logica "finestre buone" per gli alert wing foil.

Le ore arrivano come dict {"time": "YYYY-MM-DDTHH:MM", "speed": kn,
"gust": kn, "dir": gradi}, ordinate. Una finestra buona è una run di ore
CONSECUTIVE, dentro la fascia [day_start, day_end), con velocità >= soglia
e direzione in uno dei settori scelti, lunga almeno min_hours.
`end` è l'inizio dell'ultima ora buona + 1h (14-18 = 4 ore buone).
"""

SECTORS = ["N", "NE", "E", "SE", "S", "SW", "W", "NW"]


def deg_to_sector(deg):
    return SECTORS[int(((deg % 360) + 22.5) // 45) % 8]


def _hour_ok(h, rules):
    t = int(h["time"][11:13])
    return (rules["day_start"] <= t < rules["day_end"]
            and h["speed"] >= rules["min_knots"]
            and deg_to_sector(h["dir"]) in rules["sectors"])


def find_windows(hours, rules):
    windows, run = [], []

    def flush():
        if len(run) >= rules["min_hours"]:
            end = run[-1]["time"]
            end_iso = end[:11] + f"{int(end[11:13]) + 1:02d}" + end[13:]
            windows.append({
                "start": run[0]["time"],
                "end": end_iso,
                "min_speed": min(h["speed"] for h in run),
                "max_speed": max(h["speed"] for h in run),
                "sectors": sorted({deg_to_sector(h["dir"]) for h in run},
                                  key=SECTORS.index),
            })
        run.clear()

    prev = None
    for h in hours:
        contiguous = (prev is not None
                      and h["time"][:10] == prev[:10]
                      and int(h["time"][11:13]) == int(prev[11:13]) + 1)
        if _hour_ok(h, rules):
            if run and not contiguous:
                flush()
            run.append(h)
        else:
            flush()
        prev = h["time"]
    flush()
    return windows

File: windlogic.js
// Porting 1:1 di scripts/windlogic.py — stesso contratto, stessi casi di test.
const WindLogic = (() => {
  const SECTORS = ["N", "NE", "E", "SE", "S", "SW", "W", "NW"];

  function degToSector(deg) {
    return SECTORS[Math.floor((((deg % 360) + 360) % 360 + 22.5) / 45) % 8];
  }

  function hourOk(h, rules) {
    const t = parseInt(h.time.slice(11, 13), 10);
    return rules.day_start <= t && t < rules.day_end
      && h.speed >= rules.min_knots
      && rules.sectors.includes(degToSector(h.dir));
  }

  function findWindows(hours, rules) {
    const windows = [];
    let run = [];
    const flush = () => {
      if (run.length >= rules.min_hours) {
        const end = run[run.length - 1].time;
        const endIso = end.slice(0, 11)
          + String(parseInt(end.slice(11, 13), 10) + 1).padStart(2, "0")
          + end.slice(13);
        const sectors = [...new Set(run.map(h => degToSector(h.dir)))]
          .sort((a, b) => SECTORS.indexOf(a) - SECTORS.indexOf(b));
        windows.push({
          start: run[0].time,
          end: endIso,
          min_speed: Math.min(...run.map(h => h.speed)),
          max_speed: Math.max(...run.map(h => h.speed)),
          sectors,
        });
      }
      run = [];
    };
    let prev = null;
    for (const h of hours) {
      const contiguous = prev !== null
        && h.time.slice(0, 10) === prev.slice(0, 10)
        && parseInt(h.time.slice(11, 13), 10) === parseInt(prev.slice(11, 13), 10) + 1;
      if (hourOk(h, rules)) {
        if (run.length && !contiguous) flush();
        run.push(h);
      } else {
        flush();
      }
      prev = h.time;
    }
    flush();
    return { windows, hourOk };
  }

  // findWindows restituisce solo l'array nelle chiamate dell'app;
  // hourOk è esposto a parte per colorare il meteogramma.
  return {
    degToSector,
    hourOk,
    findWindows: (hours, rules) => findWindows(hours, rules).windows,
  };
})();

if (typeof module !== "undefined") module.exports = WindLogic;
